src/de/jost_net/JVerein/server/DDLTool/Updates/Update0447.java

```diff
@@ -0,0 +1,80 @@
+/**********************************************************************
+ * This program is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU General Public License as published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without 
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See 
+ * the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program. If not, 
+ * see <http://www.gnu.org/licenses/>.
+ * 
+ **********************************************************************/
+package de.jost_net.JVerein.server.DDLTool.Updates;
+
+import de.jost_net.JVerein.server.DDLTool.AbstractDDLUpdate;
+import de.jost_net.JVerein.server.DDLTool.Column;
+import de.jost_net.JVerein.server.DDLTool.Index;
+import de.willuhn.util.ApplicationException;
+import de.willuhn.util.ProgressMonitor;
+
+import java.sql.Connection;
+
+public class Update0447 extends AbstractDDLUpdate
+{
+  public Update0447(String driver, ProgressMonitor monitor, Connection conn)
+  {
+    super(driver, monitor, conn);
+  }
+
+  @Override
+  public void run() throws ApplicationException
+  {
+    execute(alterColumn("mail",
+        new Column("versand", COLTYPE.TIMESTAMP, 0, "NULL", false, false)));
+
+    execute(alterColumn("mailempfaenger",
+        new Column("versand", COLTYPE.TIMESTAMP, 0, "NULL", false, false)));
+
+    execute(dropForeignKey("fkAnfangsbestand1", "anfangsbestand"));
+
+    execute(createForeignKey("fkAnfangsbestand1", "anfangsbestand", "konto",
+        "konto", "id", "CASCADE", "RESTRICT"));
+
+    execute(dropForeignKey("fkLehrgang2", "lehrgang"));
+
+    execute(createForeignKey("fkLehrgang2", "lehrgang", "lehrgangsart",
+        "lehrgangsart", "id", "RESTRICT", "RESTRICT"));
+
+    Index idx = new Index("ixBuchung9", false);
+    Column col = new Column("splitid", COLTYPE.BIGINT, 0, null, false,
+        false);
+    idx.add(col);
+    execute(idx.getCreateIndex("buchung"));
+    
+    execute(createForeignKey("fkBuchung9", "buchung", "splitid", "buchung",
```

Comment:
Was ist der Sinn für diesen Index? Er verhindert nicht, dass man Splitbuchungen löschen kann. Er verhindert nur, dass man die Hauptbuchung nicht löschen kann wenn sie noch von Buchungen referenziert wird. Da sich die Hauptbuchung selbst referenziert kann man sie nicht mehr löschen wenn man nicht vorher seine SplitId auf null setzt.
Beim Auflösen der Splitbuchung ist das der Fall. Bei der Funktion Datenbank bereinigen für das Löschen von Buchungen ist das nicht der Fall. Dort geht das Löschen der Hauptbuchungen schief während die Gegenbuchung und die Splitbuchungen gelöscht werden.
Es muss in DbBereinigenControl Zeile 477 geprüft werden ob die Buchung eine Splitbuchung Hauptbuchung ist und dann der SplitId auf null setzen vor dem Löschen.

Um zu verhindern, dass man die Splitbuchungen nicht löschen kann müsste das im deleteCheck der Buchung eingebaut werden. Dann muss man aber im Code von Splitbuchung Auflösen und im DbBereinigen auch bei denen die SplitId vor dem Löschen auf null setzen.

Momentan wird das Löschen von Splitbuchungen aber sowieso über den Menüeintrag abgefangen und verhindert.

Reply:
Der Sinn ist, das die Hauptbuchung nicht gelöscht wird solange Splitbuchungen bestehen. Ich hatte übersehen, dass auch bei der Hauptbuchung die Splitid gesetzt ist. In dem Fall würde ich den Key auf "on Delete Casade" setzen.
Den Sinn der Foreign Keys sehe ich in erster Linie darin, dass die Konsistenz der Datenbank erhalten bleibt, unabhängig davon ob die durch die Software oder manuell geändert wird. Daher finde ich es wichtig, dass alle Felder die sich auf andere Tabellen beziehen auch als Foreign Keys angelegt sind.

Reply:
Ja, ich finde es auch gut wenn alle DB Beziehungen auch als Foreign Key modelliert sind.

Mit der neuen SplitId Beziehung auf CASCADE werden jetzt beim Löschen der Hauptbuchung auch die Splitbuchungen mit gelöscht. Das führt in DbBereinigen wieder zu Exceptions die ausgegeben werden. Ist nämlich die Haupbuchung vor einer Splitbuchung in der Löschliste gibt es die Splitbuchung nicht mehr wenn sie gelöscht werden soll. Ein ähnliches Problem gab es bei den Sollbuchungen.

Könntest du noch in DbBereinigenControl Zeile 477 das b.delete() in das darauf folgende try verschieben. Dann kommt die Exception nicht zur Ausgabe. Das wäre aber dann bei allen Buchungen abgefangen. Vielleicht sollte man es nur im try machen wenn es eine Splitbuchung ist. Oder kann man vorher abfragen, ob die Buchung noch existiert?

Reply:
Stimmt. Habe ich gemacht